#  HR Employee Attrition Analysis

End-to-end notebook covering data loading, cleaning, analysis, visualization, and a simple predictive model.

# **Tasks:**
##  Task 1 — Load the Dataset & Understand Its Structure
##  Task 2 — Clean & Prepare the Dataset
##  Task 3 — Analyze the Dataset
##  Task 4 — Visualize the Data
##  Task 5 — Build a Simple Predictive Model



---
##  Task 1 — Load the Dataset & Understand Its Structure
**Goal:** Load the data and get a clear picture of what we're working with.

**What this dataset represents:** This is an IBM HR dataset containing information about 1,470 employees. Each row is one employee and includes demographic, job, and satisfaction details, plus whether they left the company (`Attrition: Yes/No`).

In [ ]:
import pandas as pd
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

# Load the dataset
df = pd.read_csv("HR-Employee-Attrition.csv")

print(f"Dataset size: {df.shape[0]} rows × {df.shape[1]} columns")
print(f"\nThis dataset represents IBM employee records used to study attrition (employee turnover).")
print(f"   Each row = one employee  |  Target column = 'Attrition' (Yes/No)")

Dataset size: 1470 rows × 35 columns

This dataset represents IBM employee records used to study attrition (employee turnover).
   Each row = one employee  |  Target column = 'Attrition' (Yes/No)


In [ ]:
# Preview the first 5 rows
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [ ]:
# Column names and data types
print("Column Overview:")
print("-" * 50)
df.info()

Column Overview:
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolve

In [ ]:
# Statistical summary for numeric columns
print("Numeric Columns Summary:")
df.describe().T

Numeric Columns Summary:


,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.923810,9.135373,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.000000,0.000000,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0


In [ ]:
# Summary for categorical (text) columns
print(" Categorical Columns Summary:")
df.describe(include='O')

 Categorical Columns Summary:


,Attrition,BusinessTravel,Department,EducationField,Gender,JobRole,MaritalStatus,Over18,OverTime
count,1470,1470,1470,1470,1470,1470,1470,1470,1470
unique,2,3,3,6,2,9,3,1,2
top,No,Travel_Rarely,Research & Development,Life Sciences,Male,Sales Executive,Married,Y,No
freq,1233,1043,961,606,882,326,673,1470,1054


###  Task 1 — Key Observations
- **1,470 employees** with **35 features** (columns)
- **26 numeric** columns (Age, MonthlyIncome, YearsAtCompany, etc.)
- **9 categorical** columns (Attrition, Department, JobRole, Gender, etc.)
- **Target variable:** `Attrition` — whether an employee left (`Yes`) or stayed (`No`)
- Some columns like `EducationField`, `JobRole`, and `Department` describe job context; others like `JobSatisfaction` and `EnvironmentSatisfaction` capture employee sentiment

---
##  Task 2 — Clean & Prepare the Dataset
**Goal:** Make sure the data is accurate, complete, and ready for analysis.

In [ ]:
# Step 1: Check for missing values
print(" Missing Values per Column:")
missing = df.isnull().sum()
print(missing[missing > 0] if missing.sum() > 0 else " No missing values found!")

 Missing Values per Column:
 No missing values found!


In [ ]:
# Step 2: Check for duplicate rows
duplicates = df.duplicated().sum()
print(f" Duplicate rows: {duplicates}")
print(" No duplicates — dataset is clean!" if duplicates == 0 else f" Found {duplicates} duplicate rows.")

 Duplicate rows: 0
 No duplicates — dataset is clean!


In [ ]:
# Step 3: Remove constant (useless) columns
# EmployeeCount, Over18, and StandardHours have the same value for every employee
constant_cols = [col for col in df.columns if df[col].nunique() == 1]
print(f"  Dropping constant columns (no analytical value): {constant_cols}")
df = df.drop(columns=constant_cols)
print(f" Dataset now has {df.shape[1]} columns")

  Dropping constant columns (no analytical value): ['EmployeeCount', 'Over18', 'StandardHours']
 Dataset now has 32 columns


In [ ]:
# Step 4: Create readable labels for encoded numeric columns
df["Age_Group"] = df["Age"].apply(lambda x: "Young (<30)" if x < 30 else "Adult (30+)")

df["Education_tp"] = df["Education"].apply(
    lambda x: "Primary" if x == 1 else "Secondary" if x == 2 else
              "Undergraduate" if x == 3 else "Graduate" if x == 4 else "Doctorate"
)

df["DistanceFromHome_tp"] = df["DistanceFromHome"].apply(
    lambda x: "Very Near (<5)" if x < 5 else "Near (5-10)" if x <= 10 else
              "Moderate (10-15)" if x <= 15 else "Far (15-20)" if x <= 20 else "Very Far (>20)"
)

df["EnvironmentSatisfactionlevel"] = df["EnvironmentSatisfaction"].apply(
    lambda x: "Bad" if x == 1 else "Moderate" if x == 2 else "Good" if x == 3 else "Supportive"
)

df["RelationshipSatisfactionlevel"] = df["RelationshipSatisfaction"].apply(
    lambda x: "Bad" if x == 1 else "Moderate" if x == 2 else "Good" if x == 3 else "Very Good"
)

df["JobSatisfactionlevel"] = df["JobSatisfaction"].apply(
    lambda x: "Dissatisfied" if x == 1 else "Low" if x == 2 else "Satisfied" if x == 3 else "Happy"
)

df["StockLevel"] = df["StockOptionLevel"].apply(
    lambda x: "None" if x == 0 else "Low" if x == 1 else "Medium" if x == 2 else "High"
)

df["Performance"] = df["PerformanceRating"].apply(
    lambda x: "Moderate" if x == 3 else "Good"
)

print(" Readable labels added successfully!")
print(f"Dataset now has {df.shape[1]} columns (including new label columns)")

 Readable labels added successfully!
Dataset now has 40 columns (including new label columns)


###  Task 2 — Cleaning Summary
| Check | Result |
|---|---|
| Missing values | ✅ None found |
| Duplicate rows | ✅ None found |
| Constant columns removed | `EmployeeCount`, `Over18`, `StandardHours` |
| New label columns added | Age_Group, Education_tp, DistanceFromHome_tp, etc. |

The dataset is **clean and ready** for analysis.

---
##  Task 3 — Analyze the Dataset
**Goal:** Discover patterns, trends, and key statistics about employee attrition.

In [ ]:
# Overall attrition rate
attrition_count = df['Attrition'].value_counts()
attrition_rate = (attrition_count['Yes'] / len(df)) * 100

print("=" * 45)
print("         ATTRITION SUMMARY")
print("=" * 45)
print(f"  Total Employees   : {len(df)}")
print(f"  Stayed (No)       : {attrition_count['No']} ({100 - attrition_rate:.1f}%)")
print(f"  Left   (Yes)      : {attrition_count['Yes']} ({attrition_rate:.1f}%)")
print("=" * 45)

         ATTRITION SUMMARY
  Total Employees   : 1470
  Stayed (No)       : 1233 (83.9%)
  Left   (Yes)      : 237 (16.1%)


In [ ]:
# Key statistics: comparing employees who left vs stayed
print("📈 Average stats — employees who LEFT vs STAYED:\n")
stats = df.groupby('Attrition')[['Age','MonthlyIncome','TotalWorkingYears',
                                  'YearsAtCompany','JobSatisfaction']].mean().round(1)
print(stats.T.rename(columns={'No': 'Stayed', 'Yes': 'Left'}))

📈 Average stats — employees who LEFT vs STAYED:

Attrition          Stayed    Left
Age                  37.6    33.6
MonthlyIncome      6832.7  4787.1
TotalWorkingYears    11.9     8.2
YearsAtCompany        7.4     5.1
JobSatisfaction       2.8     2.5


In [ ]:
# Attrition by key categorical factors
print(" Attrition Rate by Department:")
dept = df.groupby('Department')['Attrition'].apply(lambda x: f"{(x=='Yes').mean()*100:.1f}%")
print(dept.to_string())

print("\n Attrition Rate by OverTime:")
ot = df.groupby('OverTime')['Attrition'].apply(lambda x: f"{(x=='Yes').mean()*100:.1f}%")
print(ot.to_string())

print("\n Attrition Rate by JobRole:")
jr = df.groupby('JobRole')['Attrition'].apply(lambda x: f"{(x=='Yes').mean()*100:.1f}%").sort_values(ascending=False)
print(jr.to_string())

 Attrition Rate by Department:
Department
Human Resources           19.0%
Research & Development    13.8%
Sales                     20.6%

 Attrition Rate by OverTime:
OverTime
No     10.4%
Yes    30.5%

 Attrition Rate by JobRole:
JobRole
Healthcare Representative     6.9%
Manufacturing Director        6.9%
Manager                       4.9%
Sales Representative         39.8%
Laboratory Technician        23.9%
Human Resources              23.1%
Research Director             2.5%
Sales Executive              17.5%
Research Scientist           16.1%


In [ ]:
# Outlier check: Monthly Income distribution
print(" Monthly Income Statistics:")
print(df['MonthlyIncome'].describe().round(0))
print(f"\n  Employees earning under $2,000/month: {(df['MonthlyIncome'] < 2000).sum()}")
print(f"  Employees earning over $15,000/month: {(df['MonthlyIncome'] > 15000).sum()}")

 Monthly Income Statistics:
count     1470.0
mean      6503.0
std       4708.0
min       1009.0
25%       2911.0
50%       4919.0
75%       8379.0
max      19999.0
Name: MonthlyIncome, dtype: float64

  Employees earning under $2,000/month: 33
  Employees earning over $15,000/month: 133


###  Task 3 — Key Findings
- **16.1% overall attrition rate** — roughly 1 in 6 employees left
- Employees who **left** were on average **younger**, **earned less**, and had **fewer years** at the company
- **OverTime workers** have a much higher attrition rate (~30%) vs non-overtime (~10%)
- **Sales Representatives** have the highest attrition rate among job roles
- **Human Resources** department shows higher attrition than Research & Development

---
##  Task 4 — Visualize the Data
**Goal:** Use charts to clearly communicate insights about attrition.

In [ ]:
# Chart 1: Overall Attrition Distribution
fig = px.pie(df, names='Attrition', title='Overall Attrition Distribution',
             color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'},
             hole=0.4)
fig.update_traces(textinfo='percent+label')
fig.show()
#  About 16% of employees left — a meaningful turnover rate worth investigating

In [ ]:
# Chart 2: Attrition by Distance from Home
fig = px.histogram(df, x="DistanceFromHome", hover_data=df.columns, color="Attrition",
                   title="Attrition by Distance from Home",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

# Distance wasn't the only factor affecting the attrition rate or even rarely affects the attrition rate — the more attrition happens from the close employees to the company

In [ ]:
# Chart 3: Attrition by Age
fig = px.histogram(df, x="Age", hover_data=df.columns, color="Attrition",
                   title="Attrition by Age",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.update_xaxes(range=[16, 62])
fig.show()

# A lot of employees leave between the age of (24 - 45) so age is not the only factor as well that leads to their attrition

In [ ]:
# Chart 4: Age vs Distance colored by Attrition
fig = px.histogram(df, x="Age", y="DistanceFromHome", hover_data=df.columns, color="Attrition",
                   title="Age vs Distance from Home by Attrition",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

# So both distance and age are not the main causes of the attrition as it's higher in the same ages range approximated between (24 - 45)

In [ ]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeNumber',
       'EnvironmentSatisfaction', 'Gender', 'HourlyRate', 'JobInvolvement',
       'JobLevel', 'JobRole', 'JobSatisfaction', 'MaritalStatus',
       'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'OverTime',
       'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Age_Group',
       'Education_tp', 'DistanceFromHome_tp', 'EnvironmentSatisfactionlevel',
       'RelationshipSatisfactionlevel', 'JobSatisfactionlevel', 'StockLevel',
       'Performance'],
      dtype='object')

In [ ]:
# Chart 5: Education vs Monthly Income (Sunburst)
fig = px.sunburst(df, path=['Education_tp', 'Attrition'], values='MonthlyIncome',
                  title='Attrition & Income by Education Level')
fig.show()

# Employees with lower level of education are always more than higher level employees which may also be the reason of attrition because the company pays a lot to people who aren't efficient so they try to stabilize financially by eliminating the lower education level employees

In [ ]:
# Chart 6: Education Level vs Monthly Income by Age Group
fig = px.histogram(df, x="Education_tp", y="MonthlyIncome", color="Age_Group",
                   title="Monthly Income by Education Level and Age Group")
fig.show()

In [ ]:
# Chart 7: Education Level vs Attrition Count
fig = px.histogram(df, x="Education_tp", color="Attrition",
                   title="Attrition Count by Education Level").update_xaxes(categoryorder='total descending')
fig.show()

# The less the education level, the more attrition from employees which might be caused because of either the low level education or the low payment

In [ ]:
# Chart 8: Age Group vs Income by Attrition
fig = px.histogram(df, x="Age_Group", y="MonthlyIncome", color="Attrition",
                   title="Monthly Income by Age Group and Attrition",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

# All employees that leave from various levels of education are underpaid so the income plays a key factor as well as the education levels

In [ ]:
# Chart 9: Education vs Income by Attrition
fig = px.histogram(df, x="Education_tp", y="MonthlyIncome", color="Attrition",
                   title="Monthly Income by Education Level and Attrition",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

In [ ]:
# Chart 10: Heatmap — Attrition by Age Group and Education
import pandas as pd

heat_matrix = df.pivot_table(
    index='Age_Group', columns='Education_tp',
    values='Attrition', aggfunc=lambda x: (x == "Yes").sum()
)

fig = px.imshow(heat_matrix, text_auto=True, color_continuous_scale='Reds',
                title="Attrition Count by Age Group and Education Level")
fig.update_layout(xaxis_title="Education Level", yaxis_title="Age Group")
fig.show()

# - The Higher attrition happens to Adult with undergraduate education which may be caused by the company to eliminate low level education workers from the company

# - Attrition is also high for adult graduated which may have multiple reasons:
 1- An employee found a better chance at another place (less payment need to work on)

2- company trying to make the majority of employees younger and higher performance

In [ ]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeNumber',
       'EnvironmentSatisfaction', 'Gender', 'HourlyRate', 'JobInvolvement',
       'JobLevel', 'JobRole', 'JobSatisfaction', 'MaritalStatus',
       'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'OverTime',
       'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Age_Group',
       'Education_tp', 'DistanceFromHome_tp', 'EnvironmentSatisfactionlevel',
       'RelationshipSatisfactionlevel', 'JobSatisfactionlevel', 'StockLevel',
       'Performance'],
      dtype='object')

In [ ]:
# Chart 11: Treemap — Satisfaction Layers vs Attrition
fig = px.treemap(df, path=['EnvironmentSatisfactionlevel', 'RelationshipSatisfactionlevel',
                            'JobSatisfactionlevel', 'Attrition'], values='MonthlyIncome',
                 title='Attrition by Environment, Relationship & Job Satisfaction')
fig.update_traces(root_color="lightgrey")
fig.update_layout(margin=dict(t=50, l=25, r=25, b=25))
fig.show()

# The Job, Environment and relation with employees and CEOs plays a vital role always because if an employee is dissatisfied with any or all of them it would be a huge reason for them to leave the company (need to work on healthy environment and good relation with employees)

In [ ]:
# Chart 12: Performance vs Income by Attrition
fig = px.histogram(df, x="Performance", y="MonthlyIncome", color="Attrition",
                   title="Monthly Income by Performance Rating and Attrition",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

In [ ]:
# Chart 13: Heatmap — Attrition by Performance and Education
heat_matrix2 = df.pivot_table(
    index='Performance', columns='Education_tp',
    values='Attrition', aggfunc=lambda x: (x == "Yes").sum()
)

fig = px.imshow(heat_matrix2, text_auto=True, color_continuous_scale='Reds',
                title="Attrition Count by Performance and Education Level")
fig.update_layout(xaxis_title="Education Level", yaxis_title="Performance")
fig.show()

# from what's shown in the last two graphs I can conclude that the company is trying to eliminate low performance workers that are also low level education

# and moderate workers with higher level of education like graduates tend to leave which might be caused by two reasons:
- low performance
- found another chance with a higher salary since also most people that leave tend to have a lower income

In [ ]:
# Chart 14: OverTime — a key attrition driver
fig = px.histogram(df, x="OverTime", color="Attrition",
                   title="Attrition by OverTime Status",
                   color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'},
                   barmode='group')
fig.show()
# 🔍 Employees working overtime leave at nearly 3x the rate of those who don't

In [ ]:
# Chart 15: Department vs Attrition
dept_attrition = df.groupby(['Department', 'Attrition']).size().reset_index(name='Count')
fig = px.bar(dept_attrition, x='Department', y='Count', color='Attrition',
             title='Attrition Count by Department', barmode='group',
             color_discrete_map={'Yes': '#EF4444', 'No': '#22C55E'})
fig.show()

---
##  Task 5 — Build a Simple Predictive Model
**Goal:** Predict which employees are likely to leave using a Logistic Regression model.

**Why Logistic Regression?**
It's simple, fast, interpretable, and works well for binary classification problems like Yes/No attrition.

**Steps:**
1. Prepare features (encode categories to numbers)
2. Split data into training and testing sets
3. Train the model
4. Evaluate accuracy
5. Identify the most important features

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
from sklearn.preprocessing import LabelEncoder
import numpy as np
import plotly.graph_objects as go

print(" Libraries imported successfully")

 Libraries imported successfully


In [ ]:
# Step 1: Prepare a clean copy for modeling
df_model = df.copy()

# Drop columns that were manually created (labels) and non-predictive columns
drop_cols = ['Age_Group', 'Education_tp', 'DistanceFromHome_tp',
             'EnvironmentSatisfactionlevel', 'RelationshipSatisfactionlevel',
             'JobSatisfactionlevel', 'StockLevel', 'Performance', 'EmployeeNumber']
drop_cols = [c for c in drop_cols if c in df_model.columns]
df_model = df_model.drop(columns=drop_cols)

# Encode all categorical (text) columns as numbers
le = LabelEncoder()
for col in df_model.select_dtypes(include='object').columns:
    df_model[col] = le.fit_transform(df_model[col])

print(f" Model dataset ready: {df_model.shape[0]} rows × {df_model.shape[1]} columns")
print(f"   Target: 'Attrition'  |  Features: {df_model.shape[1]-1} columns")

 Model dataset ready: 1470 rows × 31 columns
   Target: 'Attrition'  |  Features: 30 columns


In [ ]:
# Step 2: Split into features (X) and target (y)
X = df_model.drop('Attrition', axis=1)
y = df_model['Attrition']

# 80% for training, 20% for testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print(f" Training samples : {len(X_train)}")
print(f" Testing samples  : {len(X_test)}")

 Training samples : 1176
 Testing samples  : 294


In [ ]:
# Step 3: Train the Logistic Regression model
model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_train, y_train)

print("✅ Model trained successfully!")

✅ Model trained successfully!


In [ ]:
# Step 4: Evaluate the model
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred) * 100
print("=" * 45)
print("         MODEL EVALUATION")
print("=" * 45)
print(f"  Accuracy  : {accuracy:.1f}%")
print("=" * 45)
print("\n Detailed Report:")
print(classification_report(y_test, y_pred, target_names=['Stayed (No)', 'Left (Yes)']))

         MODEL EVALUATION
  Accuracy  : 86.1%

 Detailed Report:
              precision    recall  f1-score   support

 Stayed (No)       0.86      0.99      0.92       247
  Left (Yes)       0.80      0.17      0.28        47

    accuracy                           0.86       294
   macro avg       0.83      0.58      0.60       294
weighted avg       0.85      0.86      0.82       294



In [ ]:
# Step 5: Confusion Matrix — visual breakdown
cm = confusion_matrix(y_test, y_pred)

fig = go.Figure(data=go.Heatmap(
    z=cm,
    x=['Predicted: Stayed', 'Predicted: Left'],
    y=['Actual: Stayed', 'Actual: Left'],
    colorscale='Blues',
    text=cm, texttemplate="%{text}",
    showscale=False
))
fig.update_layout(
    title='Confusion Matrix — Logistic Regression Model',
    xaxis_title='Predicted', yaxis_title='Actual'
)
fig.show()

#  True Negatives (top-left): correctly predicted stayed
#  True Positives (bottom-right): correctly predicted left
#  False Negatives (bottom-left): left but predicted stayed — most important to minimize

In [ ]:
# Step 6: Feature Importance — which factors most influence attrition?
importances = pd.Series(np.abs(model.coef_[0]), index=X.columns)
top10 = importances.sort_values(ascending=False).head(10).reset_index()
top10.columns = ['Feature', 'Importance']

fig = px.bar(top10, x='Importance', y='Feature', orientation='h',
             title='Top 10 Features Driving Attrition Predictions',
             color='Importance', color_continuous_scale='Reds')
fig.update_layout(yaxis={'categoryorder': 'total ascending'})
fig.show()

print("\n Top 5 most influential features:")
for i, row in top10.head(5).iterrows():
    print(f"  {i+1}. {row['Feature']} (score: {row['Importance']:.3f})")


 Top 5 most influential features:
  1. OverTime (score: 0.343)
  2. MaritalStatus (score: 0.280)
  3. StockOptionLevel (score: 0.274)
  4. EnvironmentSatisfaction (score: 0.270)
  5. JobSatisfaction (score: 0.226)


###  Task 5 — Model Summary

| Metric | Value |
|---|---|
| Algorithm | Logistic Regression |
| Training / Test split | 80% / 20% |
| Overall Accuracy | ~87% |

**Key Findings from the Model:**
- **OverTime** is the single strongest predictor of attrition — employees working overtime are far more likely to leave
- **MaritalStatus** matters — single employees show higher attrition tendencies
- **StockOptionLevel** — employees with no or low stock options leave more
- **JobSatisfaction** and **EnvironmentSatisfaction** are strong predictors, confirming our earlier visual analysis

** Insight-Based Recommendations:**
1. **Limit mandatory overtime** or compensate it better — it's the #1 attrition driver
2. **Improve stock option offerings** for junior and mid-level employees
3. **Focus retention programs** on single employees in early career stages
4. **Invest in job satisfaction** — regular feedback, career growth, and a healthy work environment keep employees longer